contracts/EnhancedStepFunctions.sol

```diff
@@ -160,61 +162,40 @@ contract ConversionRates2 is ConversionRates {
         return getImbalance(token, rateUpdateBlock, currentBlock);
     }
 
-    function executeStepFunction(StepFunction f, int x) internal pure returns(int) {
-        uint len = f.y.length;
-        if (len == 0) { return 0; }
+    function executeStepFunction(StepFunction f, int from, int to) internal pure returns(int) {
+        if (f.y.length == 0) { return 0; }
+
+        uint len = f.x.length;
         uint ind;
-        if (x == 0) {
-            // fallback to old logics
-            for(ind = 0; ind < len; ind++) {
-                if (x <= f.x[ind]) { return f.y[ind]; }
-            }
-            return f.y[len - 1];
+
+        if (from == to) {
+            // should only happen when trade amount = 0
+            return 0;
         }
 
-        int lastStepAmount = 0; // amount from last step to compute amount to be applied bps for next step
+        int qty = to - from;
         int change = 0; // amount change from initial amount when applying bps for each step
-        int fx;
-
-        if (x > 0) {
-            // handle positive qty
-            for (ind = 0; ind < len; ind++) {
-                fx = f.x[ind];
```

Comment:
here I assume every two steps need to query values from new DB.
so we save sload operations.